### PR TITLE
Add support for 'flattened object' fields.

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR merges the `object-fields` feature branch. All commits on the branch have been individually code reviewed as part of earlier PRs.

Before merging, there are a few open issues to resolve:
- [x] The field type is currently marked 'experimental'. I've started an internal discussion to see if we can remove this tag, since the feature is useful in its current form and we don't expect huge changes in its API.
- [x] There is some chance we want to revise the type name again -- I reopened an internal issue to ask for feedback.
- [x] I will push a commit with some tweaks to the documentation (I've gained more insight on the field type from performance profiling and discussions with the Kibana team).

Original issue: #25312
Meta-issue tracking design + implementation: #33003

### CODE REVIEW

**Review by jpountz — COMMENTED** [review]
I really like how queries and aggregations work as if fields had been mapped on their own. However, this is not the case for stored fields, which makes me wonder whether we should leave it unsupported for now.

File: docs/reference/mapping/types.asciidoc [hunk 1]
```diff
@@ -82,6 +84,8 @@ include::types/date.asciidoc[]
 
 include::types/date_nanos.asciidoc[]
 
+include::types/embedded-json.asciidoc[]
+
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types.asciidoc]: 👍

File: docs/reference/mapping/types/embedded-json.asciidoc [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+be searched through simple keyword-style queries.
+
+This data type can be useful for indexing objects with a very large number of
+distinct keys. Compared to mapping each field separately, embedded JSON fields have
+the following advantages:
+
+- Only one field mapping is created for the whole object, which can help
+  prevent a <<mapping-limit-settings, mappings explosion>> due to a large
+  number of field mappings.
+- An embedded JSON field may take up less space in the index, as only one underlying
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types/embedded-json.asciidoc]: +1 I actually have a TODO to rework this docs page, I will ping you for another look when that's done.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java [hunk 3]
```diff
@@ -83,14 +100,89 @@ public FieldTypeLookup copyAndAddAll(String type,
...
+    /**
+     * Computes the total depth of this field by counting the number of parent fields
+     * in its path. As an example, the field 'parent1.parent2.field' has depth 3.
+     */
+    private static int fieldDepth(String field) {
+        int numDots = 0;
+        for (int i = 0; i < field.length(); ++i) {
+            if (field.charAt(i) == '.') {
+                numDots++;
+            }
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java]: 👍

File: docs/reference/mapping/types.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types.asciidoc]: I think these includes are just alphabetized. For the actual links to individual field types, I put it under "Specialised datatypes" to encourage users to think through whether it's appropriate for their data.

File: server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+ * (see {@link JsonFieldParser#SEPARATOR}).
+ *
+ * When 'store' is enabled, a single stored field is added containing the entire JSON object in
+ * pretty-printed format.
+ */
+public final class JsonFieldMapper extends FieldMapper {
+
+    public static final String CONTENT_TYPE = "embedded_json";
+    public static final NamedAnalyzer WHITESPACE_ANALYZER = new NamedAnalyzer(
+        "whitespace", AnalyzerScope.INDEX, new WhitespaceAnalyzer());
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java]: That makes sense -- I'll move the construction of this analyzer to `JsonFieldMapper.Builder#build` to match what we do in `KeywordFieldMapper`.

File: docs/reference/mapping/types/embedded-json.asciidoc [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+`range`, treat the values as string keywords. Highlighting is not supported on
+`json` fields.
+
+It is possible to sort on an embedded JSON field, as well as perform simple
+keyword-style aggregations such as `terms`. As with queries, there is no
+special support for numerics -- all values in the JSON object are treated as
+keywords. When sorting, this implies that values are compared lexicographically.
+
+Finally, because of the way leaf values are stored in the index, the null
+character `\0` is not allowed to appear in the keys of the JSON object.
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types/embedded-json.asciidoc]: Thanks, I had forgotten these weren't allowed by default.

**Comment by jtibshirani** [comment]
> However, this is not the case for stored fields, which makes me wonder whether we should leave it unsupported for now.

Looking at this again, I also find the behavior around stored fields to be a bit unintuitive:
- Only the root field is stored, it is not possible to load the keyed fields through `stored_fields`.
- We store the whole JSON input, which differs from what is indexed and stored in docvalues.
- We don't preserve the original formatting because we parse the JSON block, then reconstruct it to create the stored field.

I guess the main use case would be if a user wanted to retrieve the field in a search, but that using source filtering would be too expensive. I would be okay leaving it unsupported for now because the design is not that clean. @colings86 and @romseygeek checking since you reviewed this earlier -- what do you think about removing support for stored fields?

**Comment by colings86** [comment]
I’m fine with removing support for stored field until we have a more
intuitive way of exposing it

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java [hunk 6]
```diff
@@ -36,15 +38,24 @@
     final CopyOnWriteHashMap<String, MappedFieldType> fullNameToFieldType;
     private final CopyOnWriteHashMap<String, String> aliasToConcreteName;
 
+    private final CopyOnWriteHashMap<String, JsonFieldMapper> fullNameToJsonMapper;
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java]: I had actually tried this, but found it cleaner to use `JsonFieldMapper` here compared to the other option, where we use `RootJsonFieldType` and create the `KeyedJsonFieldType` objects using it. I found it nice that `JsonFieldMapper` contained consistent pairs of methods (`fieldType()` and `keyedFieldType(), `name()` and `keyedFieldName()`). To me it's acting in its role as 'field type provider'.

**Comment by jtibshirani** [comment]
@jpountz this is ready for another look. I ended up force-pushing -- the first new commit since you reviewed is d1f8ad2.

**Comment by jtibshirani** [comment]
@jpountz @colings86 this is ready for another look. The last commit since you reviewed is 7237b2b0967bf710cc033b8f40e6e082227691c4 ('Remove the experimental tag.')

File: server/src/main/java/org/elasticsearch/index/mapper/FlatObjectFieldParser.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A helper class for {@link FlatObjectFieldMapper} parses a JSON object
+ * and produces a pair of indexable fields for each leaf value.
+ */
+public class FlatObjectFieldParser {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/FlatObjectFieldParser.java]: can it be pkg-private?

File: server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java [hunk 8]
```diff
@@ -1757,5 +1757,5 @@ public void testFieldAliasesForMetaFields() throws Exception {
 
         DocumentField field = hit.getFields().get("id-alias");
         assertThat(field.getValue().toString(), equalTo("1"));
-   }
+    }
```
Comment (jpountz) [inline server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java]: maybe undo changes to this file since they are unrelated

**Review by jpountz — APPROVED** [review]
Apart from minor comments, it looks good to me.

File: docs/reference/mapping/types/flattened.asciidoc [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    Which scoring algorithm or _similarity_ should be used. Defaults
+    to `BM25`.
+
+`split_queries_on_whitespace`::
+
+    Whether <<full-text-queries,full text queries>> should split the input on
+    whitespace when building a query for this field. Accepts `true` or `false`
+    (default).
+
+<<mapping-store,`store`>>::
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types/flattened.asciidoc]: Oops, thanks I missed this.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/160_exists_query.yml [hunk 10]
```diff
@@ -1328,3 +1328,57 @@ setup:
...
+  - skip:
+      version: " - 7.99.99"
+      reason: "Flat object fields are currently only implemented in 8.0."
+
+  - do:
+      indices.create:
+        index:  flat_object_test
+        body:
+          mappings:
+            dynamic: false
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/160_exists_query.yml]: 👍 removed.

File: server/src/main/java/org/elasticsearch/index/mapper/FlatObjectFieldParser.java (hunk 7 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FlatObjectFieldParser.java]: 👍

File: server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java (hunk 8 above)
Comment (jtibshirani) [reply server/src/test/java/org/elasticsearch/search/query/SearchQueryIT.java]: 👍

**Comment by jtibshirani** [comment]
@elasticmachine run elasticsearch-ci/default-distro 
@elasticmachine run elasticsearch-ci/bwc
